curvefs/src/client/rpcclient/cli2_cient.cpp

```diff
@@ -0,0 +1,136 @@
+        LOG(WARNING) << "GetLeader failed from " << cntl.remote_side()
+                     << ", logicpool id = " << taskContext.poolID
+                     << ", copyset id = " << taskContext.copysetID
+                     << ", error = " << cntl.ErrorText();
+    } else {
+        LOG(INFO) << "GetLeader returned from " << cntl.remote_side()
+                  << ", logicpool id = " << taskContext.poolID
+                  << ", copyset id = " << taskContext.copysetID
+                  << ", leader = " << response.DebugString();
+        success = true;
+    }
+
+    excutor->NotifyRpcFinish(success);
+}
+
+#define RPCFunc [&](brpc::Channel * channel) -> int  // NOLINT
+
+bool Cli2ClientImpl::GetLeader(LogicPoolID poolID, CopysetID copysetID,
+                               const PeerInfoList &peerInfoList,
+                               int16_t currentLeaderIndex, PeerAddr *peerAddr,
+                               MetaserverID *metaserverID) {
+    int16_t index = -1;
+    peerAddr->Reset();
+
+    // std::unordered_set<std::string> peerAddrs;
+    bool getLeaderOK = false;
+    for_each(
+        peerInfoList.begin(), peerInfoList.end(),
+        [&](const CopysetPeerInfo<MetaserverID> &info) {
+            ++index;
+            // skip current index
+            if (index != currentLeaderIndex && !getLeaderOK) {
+                // make request
+                auto excutor = std::make_shared<TaskExcutor2>();
+                Cli2Closure *done = new Cli2Closure(
+                    Cli2TaskContext(
+                        poolID, copysetID,
+                        butil::endpoint2str(info.externalAddr.addr_).c_str()),
+                    excutor);
+                done->cntl.set_timeout_ms(opt_.rpcTimeoutMs);
+
+                auto task = RPCFunc {
+                    curvefs::metaserver::copyset::GetLeaderRequest2 request;
+                    request.set_poolid(poolID);
+                    request.set_copysetid(copysetID);
+
+                    curvefs::metaserver::copyset::CliService2_Stub stub(
+                        channel);
+                    stub.GetLeader(&done->cntl, &request, &done->response,
+                                   done);
+                };
+
+                // analyse response
+                if (true == excutor->DoRPCTaskAndWait(
+                                task, done->taskContext.peerAddr)) {
+                    getLeaderOK = true;
+
+                    bool has_address = done->response.peer().has_address();
+                    if (has_address) {
+                        peerAddr->Parse(done->response.peer().address());
+                    }
+
+                    bool has_id = done->response.peer().has_id();
+                    if (has_id) {
+                        *metaserverID = done->response.peer().id();
+                    }
+                }
+            }
+        });
+
+    return getLeaderOK;
+}
+
+bool TaskExcutor2::DoRPCTaskAndWait(Task2 task, std::string peerAddr) {
```

Comment:
treat `std::function` as `std::string`, it will allocate memory, so it better to pass it as const reference.

BTW, second parameter also should pass by const reference

Reply:
fix